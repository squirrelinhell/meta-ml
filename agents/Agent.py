
class Agent:
    def action_batch(self, o_batch):
        raise NotImplementedError("action_batch")
