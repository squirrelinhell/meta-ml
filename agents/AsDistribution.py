
import mandalka

from . import Agent

@mandalka.node
class AsDistribution(Agent):
    def __init__(self, agent):
        import numpy as np

        def action_batch(o_batch):
            a_batch = agent.action_batch(o_batch)

            a_batch = np.array(a_batch)
            for i, a in enumerate(a_batch):
                # Safely calculate softmax
                a = np.exp(a - np.amax(a))
                a_batch[i] = a / a.sum()

            return a_batch

        self.action_batch = action_batch
