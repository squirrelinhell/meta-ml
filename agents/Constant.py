
import mandalka

from . import Agent

@mandalka.node
class Constant(Agent):
    def __init__(self, world, seed, value):
        assert seed == 0
        del seed

        import numpy as np

        value = np.zeros(world.act_shape) + value
        assert value.shape == world.act_shape

        def step(sta_batch, obs_batch):
            tile_dim = [len(obs_batch)] + [1] * len(value.shape)
            return sta_batch, np.tile(value, tile_dim)

        self.step = step
