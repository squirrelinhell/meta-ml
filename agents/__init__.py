
class Agent:
    def action_batch(self, o_batch):
        raise NotImplementedError("action_batch")

from .RandomChoice import RandomChoice
from .Repeat import Repeat
from .SampleFromPolicy import SampleFromPolicy
from .Softmax import Softmax
