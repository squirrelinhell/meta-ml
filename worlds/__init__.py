
class World:
    def get_observation_shape(self): # -> shape [tuple of int] or None
        return None

    def get_action_shape(self): # -> shape [tuple of int]
        raise NotImplementedError("get_action_shape")

    def get_reward_shape(self): # -> shape [tuple of int]
        raise NotImplementedError("get_reward_shape")

    def after_episode(self, agent, seed): # -> (world, experience)
        # Iterating over experience should return tuples:
        # (observation, action, reward)
        raise NotImplementedError("start_episode")

    # For convenience only
    def __getattr__(self, name):
        if name in ("o_shape", "observation_shape"):
            return self.get_observation_shape()
        if name in ("a_shape", "action_shape"):
            return self.get_action_shape()
        if name in ("r_shape", "reward_shape"):
            return self.get_reward_shape()
        return object.__getattribute__(self, name)

from .Accuracy import Accuracy
from .Distribution import Distribution
from .Gym import Gym
from .Mnist import Mnist
from .PolicyNet import PolicyNet
from .Reinforce import Reinforce
