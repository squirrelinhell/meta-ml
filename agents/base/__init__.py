
from .Agent import Agent
from .StatelessAgent import StatelessAgent
from .WrapperAgent import WrapperAgent
