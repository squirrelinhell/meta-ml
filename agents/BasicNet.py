
import mandalka

from . import Agent
from worlds import World

@mandalka.node
class BasicNet(Agent):
    def __init__(self, world, seed, hidden_layers, params):
        import numpy as np

        # Wrap the original world with a meta world of parameters
        world = BasicNetParamsWorld(world, hidden_layers)

        # Create agent object (if needed)
        if not isinstance(params, Agent):
            params = params(world, seed)
            assert isinstance(params, Agent)

        # Get a single parameter vector for this instance
        _, (params,) = params.step([None], [None])
        params = np.asarray(params)
        params.setflags(write=False)
        assert params.shape == world.act_shape

        def step(sta_batch, obs_batch):
            # Use the world's TF session to run computations
            return sta_batch, world.action_batch(params, obs_batch)

        self.step = step
        self.get_parameters = lambda: params

@mandalka.node
class BasicNetParamsWorld(World):
    def __init__(self, world, hidden_layers):
        import tensorflow as tf
        import numpy as np

        assert world.rew_shape == world.act_shape

        params = Parameters()
        self.get_action_shape = lambda: (params.get_size(),)
        self.get_reward_shape = lambda: (params.get_size(),)

        obs_batch = tf.placeholder(
            tf.float32,
            (None,) + world.obs_shape
        )
        obs_size = np.prod(obs_batch.shape.as_list()[1:])
        act_size = np.prod(world.act_shape)

        def count_params():
            n_params = 0
            last_dim = obs_size
            for dim in hidden_layers:
                n_params += (last_dim + 1) * dim
                last_dim = dim
            n_params += (last_dim + 1) * act_size
            return n_params

        def affine(x, out_dim):
            assert len(x.shape.as_list()) == 2
            in_dim = int(x.shape[1])
            w = params.next_normal((in_dim, out_dim))
            b = params.next_normal((out_dim,))
            x = tf.matmul(x, w) + b
            return x * (1.0 / np.sqrt(in_dim + 1))

        def layers(inp_batch):
            layer = tf.reshape(obs_batch, (-1, obs_size))
            for dim in hidden_layers:
                layer = affine(layer, dim)
                layer = tf.nn.relu(layer)

            return affine(layer, act_size)

        params.alloc(count_params())
        act_batch = tf.reshape(
            layers(obs_batch),
            (-1,) + world.act_shape
        )

        # Backpropagation
        act_grad = tf.placeholder(
            tf.float32,
            act_batch.shape
        )
        intermediate_reward = tf.reduce_sum(tf.reduce_mean(
            tf.multiply(act_grad, act_batch),
            axis=0
        ))
        params_grad = tf.gradients(
            intermediate_reward,
            params.get_tensor()
        )[0]

        sess = tf.Session()

        def action_batch(params_value, feed_obs_batch):
            return sess.run(
                act_batch,
                feed_dict={
                    params.get_tensor(): params_value,
                    obs_batch: feed_obs_batch
                }
            )

        def trajectory(outer_agent, seed):
            assert isinstance(outer_agent, Agent)

            # Get parameters from the outer agent, and build
            # an agent to hold them
            inner_agent = BasicNet(
                world,
                None,
                hidden_layers=hidden_layers,
                params=outer_agent,
            )
            params_value = inner_agent.get_parameters()

            # IMPORTANT: Actions in experience list are assumed
            # to be generated from the inner agent.
            # So learning only works if the underlying world
            # is strictly on-policy.
            traj = world.trajectory(inner_agent, seed)
            all_obs, _, all_rew = zip(*traj)

            # Backpropagate gradient to parameters
            grad = sess.run(
                params_grad,
                feed_dict={
                    params.get_tensor(): params_value,
                    obs_batch: all_obs,
                    act_grad: all_rew
                }
            )

            return [(None, params_value, grad)]

        self.trajectory = trajectory
        self.action_batch = action_batch

class Parameters:
    def __init__(self):
        import tensorflow as tf
        import numpy as np

        tensor = None
        size = None
        pos = 0

        def alloc(n_params):
            nonlocal tensor, size
            assert tensor is None
            tensor = tf.placeholder(
                tf.float32,
                n_params,
                name="params"
            )
            size = n_params

        def next_normal(shape):
            nonlocal pos
            l = np.prod(shape)
            ret = tensor[pos:pos+l]
            pos += l
            return tf.reshape(ret, shape)

        def get_size():
            assert tensor is not None
            if pos != size:
                raise ValueError(
                    "Model used %d instead of %d parameters"
                        % (pos, size)
                )
            return size

        self.alloc = alloc
        self.next_normal = next_normal
        self.get_size = get_size
        self.get_tensor = lambda: (get_size(), tensor)[1]
