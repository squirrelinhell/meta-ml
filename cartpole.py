#!/usr/bin/env python3

import numpy as np
import tqdm

from worlds import Gym, Reinforce, PolicyNet

import debug

def train_policy(problem):
    world = PolicyNet(
        Reinforce(problem),
        normalize_rewards=True,
        batch_size=20
    )
    ep = world.start_episode(123)

    reward = 0.01 * ep.step(np.zeros(world.a_shape))
    for _, ep_len in zip(range(50), ep):
        ep_len = int(ep_len[0])
        print("=" * ep_len, ep_len)
        reward *= 0.99
        reward += 0.01 * ep.step(reward * -10.0) # ?? should be positive

    return ep

def score(world, policy):
    avg = 0.0
    for i in range(10):
        ep = world.start_episode(i)
        avg += 0.1 * np.sum([
            np.mean(ep.step(policy(inp)))
            for inp in ep
        ])
    return avg

world = Gym("CartPole-v1")

print("Average reward (random agent): %.5f" % score(
    world,
    lambda _: np.random.randn(*world.a_shape)
))

for _ in range(10):
    agent = train_policy(world)

    print("Average reward (policy network): %.5f" % score(
        Reinforce(world, test=True),
        agent.solve
    ))
