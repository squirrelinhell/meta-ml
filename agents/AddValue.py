
import mandalka

from .base import Agent, WrapperAgent
from worlds import World
from values import Value

@Agent.builder
@mandalka.node
class AddValue(WrapperAgent):
    def __init__(self, world, seed, base, add):
        add = Value.get_float(add, world.act_shape)
        base = Agent.build(base, AddValueWorld(world, add), seed)

        def process_action(a):
            return a + add.get()

        super().__init__(base, process_action=process_action)

@mandalka.node
class AddValueWorld(World):
    def __init__(self, world, add):
        self.get_observation_shape = lambda: world.obs_shape
        self.get_action_shape = lambda: world.act_shape
        self.get_reward_shape = lambda: world.rew_shape

        def trajectories(agent, n):
            assert isinstance(agent, Agent)
            return world.trajectories(
                AddValue(world, 0, base=agent, add=add),
                n
            )

        self.trajectories = trajectories
