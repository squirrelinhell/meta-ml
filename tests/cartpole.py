
from tests._ import timer

import os
import sys
import numpy as np
np.set_printoptions(precision=3, suppress=True)

from worlds import *
from agents import *
assert timer.t() < 0.15

def score(agent):
    world = Gym("CartPole-v1")
    rew_sum = 0.0
    for _ in range(10):
        for t in world.trajectory_batch(agent, range(10)):
            for o, a, r in t:
                rew_sum += np.mean(r)
    return rew_sum / 100.0

def test1():
    agent = RandomChoice(Gym("CartPole-v1"), 123, p=0.5)
    s = score(agent)
    sys.stderr.write("Reward/episode (random agent): %.5f\n" % s)
    print("Random agent sanity check:", s >= 15.0, s <= 30.0)

def test2():
    SupervisedAgent = Softmax(
        logits=BasicNet(
            hidden_layers=[128],
            batch_size=16,
            params=GradAscent(n_steps=8, log_lr=0.8)
        )
    )
    RLAgent = WholeTrajectories(
        agent=Reinforce(policy=SupervisedAgent)
    )
    agent = RLAgent(Gym("CartPole-v1"), 123)
    s = score(agent)
    sys.stderr.write("Reward/episode (policy network): %.5f\n" % s)
    print("Policy network sanity check:", s >= 50.0)

    if "DEBUG" in os.environ:
        timer.t()
        for _ in range(5):
            Gym("CartPole-v1").render(agent)
    else:
        assert timer.t() < 9.0

test1()
test2()
