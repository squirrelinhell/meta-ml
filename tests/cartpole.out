Random agent sanity check: True True
Policy network sanity check: True
